server/src/main/java/org/opensearch/common/util/BytesRefHash.java

```diff
@@ -6,179 +6,291 @@
-        spare = new BytesRef();
+    public BytesRefHash(final long initialCapacity, final BigArrays bigArrays) {
+        this(initialCapacity, DEFAULT_LOAD_FACTOR, DEFAULT_HASHER, bigArrays);
     }
 
-    // BytesRef has a weak hashCode function so we try to improve it by rehashing using Murmur3
-    // Feel free to remove rehashing if BytesRef gets a better hash function
-    private static int rehash(int hash) {
-        return BitMixer.mix32(hash);
+    public BytesRefHash(final long initialCapacity, final float loadFactor, final BigArrays bigArrays) {
+        this(initialCapacity, loadFactor, DEFAULT_HASHER, bigArrays);
     }
 
-    /**
-     * Return the key at <code>0 &lt;= index &lt;= capacity()</code>. The result is undefined if the slot is unused.
-     * <p>Beware that the content of the {@link BytesRef} may become invalid as soon as {@link #close()} is called</p>
-     */
-    public BytesRef get(long id, BytesRef dest) {
-        final long startOffset = startOffsets.get(id);
-        final int length = (int) (startOffsets.get(id + 1) - startOffset);
-        bytes.get(startOffset, length, dest);
-        return dest;
+    public BytesRefHash(final long initialCapacity, final float loadFactor, final Hasher hasher, final BigArrays bigArrays) {
+        assert initialCapacity > 0 : "initial capacity must be greater than 0";
+        assert loadFactor > 0 && loadFactor < 1 : "load factor must be between 0 and 1";
+
+        this.loadFactor = loadFactor;
+        this.hasher = hasher;
+        this.bigArrays = bigArrays;
+
+        capacity = Numbers.nextPowerOfTwo((long) (initialCapacity / loadFactor));
+        assert capacity <= MAX_CAPACITY : "required capacity too large";
+        mask = capacity - 1;
+        size = 0;
+        grow = (long) (capacity * loadFactor);
+
+        table = bigArrays.newLongArray(capacity, false);
+        table.fill(0, capacity, -1);
+        offsets = bigArrays.newLongArray(initialCapacity + 1, false);
+        offsets.set(0, 0);
+        keys = bigArrays.newByteArray(initialCapacity * 3, false);
     }
 
     /**
-     * Get the id associated with <code>key</code>
+     * Adds the given key to the hash table and returns its ordinal.
+     * If the key exists already, it returns (-1 - ordinal).
      */
-    public long find(BytesRef key, int code) {
-        final long slot = slot(rehash(code), mask);
-        for (long index = slot;; index = nextSlot(index, mask)) {
-            final long id = id(index);
-            if (id == -1L || key.bytesEquals(get(id, spare))) {
-                return id;
+    public long add(final BytesRef key) {
+        final long hash = hasher.hash(key);
+        final long fingerprint = hash & MASK_FINGERPRINT;
+
+        for (long idx = hash & mask, value, ordinal;; idx = (idx + 1) & mask) {
+            if ((value = table.get(idx)) == -1) {
+                final long val = fingerprint | size;
+                if (size >= grow) {
```

Comment:
Given the initial value of size is set to 0, wouldn't this grow un-proportionately. Essentially, creating a table of size 64 is not being used here, should that also be set to 1 to start with?

Reply:
Size is the number of elements stored in the hash table currently. Capacity is the number of slots in the backing array of the hash table.

An empty table always starts with size 0. We modified the initial capacity of the table to 32 (64 when adjusted for load factor), so as more elements are added to it, the number of times we grow is reduced.